sycl/doc/extensions/SYCL_pipes_lowering_to_SPIRV.rst

```diff
@@ -0,0 +1,111 @@
+
+Lowering of kernel to kernel pipes to SPIR-V (non-blocking)
+===========================================================
+
+This connectivity is achieved through OpTypePipeStorage, which allows a SPIR-V
+device consumer to leverage static connectivity. An OpConstantPipeStorage
+instruction must create a single instance of OpPipeStorage for each kernel to
+kernel pipe type used by any kernel within the application.
+
+OpTypePipe objects is created from OpPipeStorage using
+OpCreatePipeFromPipeStorage. The number of OpTypePipe objects created from an
+OpPipeStorage object is an implementation detail, as are the access qualifiers
+applied to those types. For example, an implementation is free to create a
+different OpTypePipe corresponding to each read and write, with unidirectional
+access qualifiers annotated, or it can create fewer OpTypePipe objects, although
+read and write pipes must be distinct according to OpReadPipe and OpWritePipe
+rules.
+
+NOTE: The SPIR-V OpReadPipe and OpWritePipe instructions are non-blocking.
+
+Details SPIR-V representation in LLVM IR
+========================================
+
+Pipe built-ins are mangled in LLVM IR to make it SPIR-V friendly.
+As an example:
+ OpReadPipe                  -> __spirv_ReadPipe
+ OpWritePipe                 -> __spirv_WritePipe
+ OpCreatePipeFromPipeStorage -> __spirv_CreatePipeFromPipeStorage_{read|write}
+
+where 0 and 1 are representing read_only and write_only access qualifiers.
+
+More about SPIR-V representation in LLVM IR can be found under the link:
+.. _SPIRVRepresentationInLLVM.rst: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst/
+
+Example of SYCL -> LLVM IR -> SPIR-V -> LLVM-IR code transformations
+====================================================================
+Consider following SYCL device code:
+.. code:: cpp
+  pipe<class some_pipe, int, 1>::write(42, SuccessCode);
+
+After compiling this code with clang we will be given following piece of IR for
+the write pipe function call (NOTE: for this implementation clang-known
+OpenCL 2.0 pipe types are reused):
+.. code:: cpp
+  define internal spir_func void @_ZN2cl4sycl4pipeIZ4mainE9some_pipeiLi1EE5writeEiRb(i32, i8* dereferenceable(1)) #4 align 2 {
+    //...
+    %12 = call spir_func i32 @_Z17__spirv_WritePipeIiEi8ocl_pipePT_ii(%opencl.pipe_wo_t addrspace(1)* %10, i32 addrspace(4)* %11, i32 4, i32 4) #8
+    //...
+  }
+
+with following declaration:
+.. code:: cpp
+  %12 = call spir_func i32 @_Z17__spirv_WritePipeIiEi8ocl_pipePT_ii(%opencl.pipe_wo_t addrspace(1)* %10, i32 addrspace(4)* %11, i32 4, i32 4) #8
+
+Yet it's looking heavy. But SPIR-V translator will drop all of these manglings
+just making a call of SPIR-V write pipe built-in:
+.. code:: cpp
+ 7 WritePipe 51 158 156 157 52 52
+
+As this implementation is based on SPIR-V representation of OpenCL program scope
+pipes, resulting code for translation back to LLVM IR from SPIR-V is good known
```

Comment:
"is good known"?

Reply:
At least these built-ins exist in clang :)
I will remove that.